File: BambooBank-Spring-Java/src/main/java/BambooBank_Spring_Java/services/EmployeeService.java
package BambooBank_Spring_Java.services;

import java.util.List;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import BambooBank_Spring_Java.exceptions.EmployeeNotFoundException;
import BambooBank_Spring_Java.pojos.Employee;
import BambooBank_Spring_Java.repositories.EmployeeRepository;






@Service
public class EmployeeService {

    private EmployeeRepository employeeRepository;

    @Autowired
    public EmployeeService(EmployeeRepository employeeRepository) {
        this.employeeRepository = employeeRepository;
    }

    public List<Employee> getAllEmployees() {
        return employeeRepository.findAll();
    }

    public Employee getEmployeeById(long id) throws EmployeeNotFoundException {
        Optional<Employee> optionalEmployee = employeeRepository.findById(id);
        return optionalEmployee.orElseThrow(() -> new EmployeeNotFoundException("Employee not found with id: " + id));
    }

    public Employee createEmployee(Employee employee) {
        return employeeRepository.save(employee);
    }

    public Employee updateEmployee(long id, Employee employee) throws EmployeeNotFoundException {
        if (!employeeRepository.existsById(id)) {
            throw new EmployeeNotFoundException("Employee not found with id: " + id);
        }
        employee.setId(id);
        return employeeRepository.save(employee);
    }

    public void deleteEmployee(long id) throws EmployeeNotFoundException {
        if (!employeeRepository.existsById(id)) {
            throw new EmployeeNotFoundException("Employee not found with id: " + id);
        }
        employeeRepository.deleteById(id);
    }



     
}

File: BambooBank-Spring-Kotlin/src/main/kotlin/bamboobank_spring/bamboobankspringkotlin/services/EmployeeService.kt
package bamboobank_spring.bamboobankspringkotlin.services


import bamboobank_spring.bamboobankspringkotlin.pojos.Employee
import bamboobank_spring.bamboobankspringkotlin.repositories.EmployeeRepository
import bamboobank_spring.bamboobankspringkotlin.exceptions.EmployeeNotFoundException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class EmployeeService @Autowired constructor(private val employeeRepository: EmployeeRepository) {

    fun getAllEmployees(): List<Employee> {
        return employeeRepository.findAll()
    }

    fun getEmployeeById(id: Long): Employee {
        return employeeRepository.findById(id)
            .orElseThrow { EmployeeNotFoundException("Employee not found with id: $id") }
    }

    fun createEmployee(employee: Employee): Employee {
        return employeeRepository.save(employee)
    }

    fun updateEmployee(id: Long, employee: Employee): Employee {
        if (!employeeRepository.existsById(id)) {
            throw EmployeeNotFoundException("Employee not found with id: $id")
        }

        // Create a new Employee instance with the updated id using the copy method
        val updatedEmployee = employee.copy(id = id)
        return employeeRepository.save(updatedEmployee)
    }


    fun deleteEmployee(id: Long) {
        if (!employeeRepository.existsById(id)) {
            throw EmployeeNotFoundException("Employee not found with id: $id")
        }
        employeeRepository.deleteById(id)
    }
}
